core/src/main/java/org/apache/brooklyn/core/entity/Entities.java

```diff
@@ -1139,6 +1141,29 @@ public static void warnOnIgnoringConfig(Entity entity, ConfigKey<?> key) {
             log.warn("Ignoring "+key+" set on "+entity+" ("+entity.getConfig(key)+")");
     }
 
+    /** Waits until the passed entity satisfies the supplied predicate. */
+    public static void waitFor(Entity entity, Predicate<Entity> condition, Duration timeout) {
+        try {
+            String description = "Waiting for " + condition + " on " + entity;
+            Tasks.setBlockingDetails(description);
+
+            Repeater repeater = Repeater.create(description)
+                .until(isSatisfied(entity, condition))
+                .limitTimeTo(timeout)
+                .backoffTo(Duration.ONE_SECOND)
+                .rethrowException();
+
+            if (!repeater.run()) {
```

Comment:
On the subject of exceptions, I toyed with the idea of throwing `TimeoutException` instead of `IllegalStateException`, but am opting for consistency/backwards compatibility with the existing `waitForServiceUp` methods. (The standard `TimeoutException` is checked.)